pybdv/bdv_datasets.py

```diff
@@ -0,0 +1,213 @@
+            print(f'position = {position}')
+
+        volume = volume[
+            source_starts[0]: source_ends[0],
+            source_starts[1]: source_ends[1],
+            source_starts[2]: source_ends[2]
+        ]
+
+    return position, volume
+
+
+def _check_shape_and_position_scaling(max_scale, position, volume,
+                                      data_path, setup_id, timepoint,
+                                      verbose=False):
+    vol_shape = np.array(volume.shape)
+    if ((position / max_scale) - (position / max_scale).astype(int)).max()\
+            or ((vol_shape / max_scale) - (vol_shape / max_scale).astype(int)).max():
+
+        # They don't scale properly:
+        # So, we have to read a volume from the target data (largest scale) that does and covers the area of where the
+        # volume belongs, which we here call target_vol with respective properties target_pos and target_shape
+
+        if verbose:
+            print('----------------------')
+            print(f'max_scale = {max_scale}')
+            print(f'position = {position}')
+            print(f'vol_shape = {vol_shape}')
+            print('----------------------')
+        target_pos = max_scale * (position / max_scale).astype(int)
+        target_shape = max_scale * np.ceil((position + vol_shape) / max_scale).astype(int) - target_pos
+        if verbose:
+            print(f'target_pos = {target_pos}')
+            print(f'target_shape = {target_shape}')
+        with open_file(data_path, mode='r') as f:
+            target_vol = f[f'setup{setup_id}/timepoint{timepoint}/s0'][
+                target_pos[0]: target_pos[0] + target_shape[0],
+                target_pos[1]: target_pos[1] + target_shape[1],
+                target_pos[2]: target_pos[2] + target_shape[2]
+            ]
+        if verbose:
+            print(f'target_vol.shape = {target_vol.shape}')
+
+        # Now we have to put the volume to this target_vol at the proper position
+        in_target_pos = position - target_pos
+        target_vol[
+            in_target_pos[0]: in_target_pos[0] + volume.shape[0],
+            in_target_pos[1]: in_target_pos[1] + volume.shape[1],
+            in_target_pos[2]: in_target_pos[2] + volume.shape[2]
+        ] = volume
+
+    else:
+
+        # Everything scales nicely, so we just have to take care that the proper variables exist
+        target_vol = volume
+        target_pos = position
+        target_shape = vol_shape
+
+    return target_vol, target_pos, target_shape
+
+
+def _scale_and_add_to_dataset(
+        data_path, setup_id, timepoint,
+        target_pos, target_vol, target_shape,
+        scales, downscale_mode, n_threads):
+
+    # Perform scaling of volume, by generating a temporary bdv file
+    # FIXME it is probably sufficient to keep it in memory, since I assume that the volume is in memory anyways, but the
+    # FIXME     make_bdv function is so very convenient right now...
```

Comment:
It should be easy to provide a `make_bdv_array` or so that just returns a list of the arrays after downscaling.
I can look into that.